### PR TITLE
qmk: 0.0.45 -> 0.0.52

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by afontaine** [comment]
Result of `nixpkgs-review pr 127863` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>python38Packages.hid</li>
    <li>python39Packages.hid</li>
    <li>qmk</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python38Packages.milc</li>
    <li>python39Packages.milc</li>
  </ul>
</details>

**Comment by bb2020** [comment]
It is good to see you are back into working on this. Frankly I did not understand what they were trying to do in the other PR.

**Comment by afontaine** [comment]
It looks like `hid` needs some extra dependencies during the check phase?

```
error: builder for '/nix/store/76zibvn0w8cj6py3qa0r64xf7cd0df9x-python3.8-hid-1.0.4.drv' failed with exit code 1;
       last 10 log lines:
       >   File "/build/hid-1.0.4/hid/__init__.py", line 30, in <module>
       >     raise ImportError(error)
       > ImportError: Unable to load any of the following libraries:libhidapi-hidraw.so libhidapi-hidraw.so.0 libhidapi-libusb.so libhidapi-libusb.so.0 libhidapi-iohidmanager.so libhidapi-iohidmanager.so.0 libhidapi.dylib hidapi.dll libhidapi-0.dll
       >
       > ----------------------------------------------------------------------
       > Ran 1 test in 0.005s
       >
       > FAILED (errors=1)
       > Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
       > error: Test failed: <unittest.runner.TextTestResult run=1 errors=1 failures=0>
```

**Comment by AndersonTorres** [comment]
>  ImportError: Unable to load any of the following libraries:libhidapi-hidraw.so libhidapi-hidraw.so.0 libhidapi-libusb.so libhidapi-libusb.so.0 libhidapi-iohidmanager.so libhidapi-iohidmanager.so.0 libhidapi.dylib hidapi.dll libhidapi-0.dll

This is the part I am not understanding. It tries to load the libraries above from `hidapi`. They were installed (just `ls /nix/store | grep hidapi` looks OK) but `hid` Python package does not find it.

https://github.com/apmorton/pyhidapi/blob/5a872c81cf8034c011f215b5bca672762df030b4/hid/__init__.py#L9

**Comment by bb2020** [comment]
Is setting `doCheck = false` an option for `hid` package? :smiley:

**Comment by AndersonTorres** [comment]
Nope. This is not a mere stress test.

It is something like `#include <stdio.h>` in a filesystem without a libc.

**Comment by afontaine** [comment]
It looks like `pyusb` [patches itself](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pyusb/default.nix#L17), as does [hidapi](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/hidapi/default.nix#L28)

**Comment by AndersonTorres** [comment]
hidapi is not the same as hid.
At least not for Python Package Index.

**Comment by afontaine** [comment]
> hidapi is not the same as hid.
> At least not for Python Package Index.

No, but they both seem to rely on libhid, as does `pyusb`.

I'm not familiar with python though, so I do not know how the library search happens. Is there an environment variable to set to include `libhid` in a search path?

File: pkgs/development/python-modules/milc/default.nix [hunk 1]
```diff
@@ -30,6 +39,9 @@ buildPythonPackage rec {
 
   pythonImportsCheck = [ "milc" ];
 
+  # checks are failing
+  doCheck = false;
```
Comment (jonringer) [inline pkgs/development/python-modules/milc/default.nix:43]: Um, do you mind expanding on the failures?

They could be actual cause for concern.

Comment (AndersonTorres) [reply pkgs/development/python-modules/milc/default.nix:43]: Yes, I will open this now

**Comment by afontaine** [comment]
@AndersonTorres I looked into this a little more and for `hid`'s check phase to pass, `${hidapi}/lib` needs to be in the `LD_LIBRARY_PATH`. else python can't find the libraries to load.

I'm not 100% sure how the right way to get it there is, and probably persist it to runtime (maybe @jonringer knows the way?)

**Comment by AndersonTorres** [comment]
@afontaine many thanks! I have worked out the script.

**Comment by AndersonTorres** [comment]
If I am not wrong yet, `qmk` CLI can do everything except compilation.

Now comes the most confusing part: how to deal with the cross compilers.

**Comment by bb2020** [comment]
I can think of two ways to solve this problem.

Packaging `qmk_firmware` and including `qmk` as a dependency for it. `qmk_firmware` should compile all of the firmwares by default, but it also should have an overridable attribute set. Keyboard files will be copied into the `nix store` and built alongside firmware files. Custom keyboard users will preferably use an overlay to build a crafted firmware and then get the `hex` file from  `nix store` to flash it. `qmk flash` is still unusable with this approach.

```
packageOverrides = super: let self = super.pkgs; in {
  myQMK = pkgs.qmk_firmware.override { 
    customLayout = true;
    customSrc = fetchFromGitHub { ... };
    customDst = "keyboads/ergodox_ez/keymaps/bb2020";
    customPreBuild = "qmk json2c layout.json >keymap.c";
    
    makeFlags = "qmk compile -kb ergodox_ez -km bb2020";
  }
}
```

Another solution would be pushing QMK's `shell.nix` into `nixpkgs` and making it work without triggering any compilation (especially `gcc`). This approach would be similar to [androidenv](https://github.com/NixOS/nixpkgs/tree/nixos-unstable/pkgs/development/mobile/androidenv/examples/shell.nix). This is also problematic because there will be inconsistency between `qmk_firmware` fork and build tools. It is because every time the fork is pulled from upstream, one should call `qmk setup` to update the build environment.

File: pkgs/tools/misc/qmk/default.nix [hunk 2]
```diff
@@ -3,38 +3,59 @@
...
-      sha256 = "0lwi1dz35p07vha5gwq2jxm5q49vm99ix4jyhd6g6ypqbq1qiwc8";
-    })
-  ];
-
   nativeBuildInputs = with python3.pkgs; [
+    flake8
+    nose2
+    pep8-naming
     setuptools-scm
+    yapf
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/qmk/default.nix:23]: Are those checkInputs? We do not need to run linting tests.

File: pkgs/development/python-modules/hid/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    hidapi=${hidapi}/lib/
+    test -d $hidapi || { echo "ERROR: $hidapi doesn't exist, please update/fix this build expression."; exit 1; }
+    sed -i -e "s|libhidapi|$hidapi/libhidapi|" hid/__init__.py
+  '';
+
+  meta = with lib; {
+    description = "hidapi bindings in ctypes";
+    homepage = "https://github.com/apmorton/pyhidapi";
+    license = with licenses; [ mit ];
+    maintainers = with maintainers; [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/hid/default.nix:32]: Missing maintainer.

File: pkgs/development/python-modules/hid/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "sha256-9hsDgvN6M0vIuoYEvIS5SHXuT1lPu6+CssOz6CeIP8E=";
+  };
+
+  propagatedBuildInputs = [ hidapi ];
+
+  checkInputs = [ nose ];
+
+ postPatch = lib.optionalString stdenv.isLinux ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/hid/default.nix:22]: Does the package even work on other platforms?

Comment (AndersonTorres) [reply pkgs/development/python-modules/hid/default.nix:22]: I can't know for sure.
The code looks for DLLs, DYLIBs and SOs by searching them directly by filename.

Comment (afontaine) [reply pkgs/development/python-modules/hid/default.nix:22]: I just tested and the patch seems to work on MacOS as well, so we can probably remove the conditional.

**Comment by AndersonTorres** [comment]
> 1. Packaging `qmk_firmware` and including `qmk` as a dependency for it. `qmk_firmware` should compile all of the firmwares by default, but it also should have an overridable attribute set.

It is unusual that the original source code would be compiled in its entirety. Even because there are entire sections of that huge tree that do not compile at all (the original nixpkgs expression removed them completely).
Further, the firmwares are efemeral: they are generated, flashed and discarded. Therefore it does not make sense so much sense to cache the results of this compilation (and I tell you, it is a big and sluggish compilation) inside `/nix/store/`. Packaging `qmk_firmware` this way is counterproductive.

Also, the README of QMK says that QMK can work with other trees besides the original QMK Firmware. Indeed, many custom keyboard and mouse distributors release their modified QMK trees without merging them directly into the original one.

For the reasons above I do not think a qmk_firmware package is a good approach.

> 
> This is also problematic because there will be inconsistency between `qmk_firmware` fork and build tools. It is because every time the fork is pulled from upstream, one should call `qmk setup` to update the build environment.

How do this affect `qmk_cli`?

**Comment by afontaine** [comment]
I agree, I don't think it makes sense to have a `qmk_firmware` package.

I think a `build.nix` script within the `qmk_firmware` project _could_ be interesting though 🤔 and that we should add this bumped `qmk_cli` as a `buildInput` for the existing `shell.nix` file.

**Comment by AndersonTorres** [comment]
A Nix file inside qmk_firmware upstream source tree? it would be great!

**Comment by bb2020** [comment]
I think it should be in `nixpkgs` because the way I see it, `shell.nix` in `qmk_firmware` is almost always outdated. Furthermore it starts to build `gcc` from scratch when it is initiated.

**Comment by bb2020** [comment]
> Also, the README of QMK says that QMK can work with other trees besides the original QMK Firmware. Indeed, many custom keyboard and mouse distributors release their modified QMK trees without merging them directly into the original one.

This can be worked out by overriding package `src`. As long as dependencies are the same, nix package can build their tree too.

> How do this affect `qmk_cli`?

It doesn't but, we still can't build the firmware with current `qmk_cli`. :smiley:

**Comment by afontaine** [comment]
> A Nix file inside qmk_firmware upstream source tree? it would be great!

There's already a `shell.nix`! what's one more? :wink:

@bb2020 maybe this would make more sense as a convenient _builder_ as opposed to a package?

Even then, though, how do you envision the workflow? I think that might help find the best path forward.

**Comment by bb2020** [comment]
> @bb2020 maybe this would make more sense as a convenient _builder_ as opposed to a package?
> 
> Even then, though, how do you envision the workflow? I think that might help find the best path forward.

Yes, it is much better. For example, a new package called `qmk-builder` that exposes a wrapper named `qmk-shell` that spawns `mkShell`. The shell includes cross compilers (`avr`, `arm`, `teensy`), python packages and `qmk-cli` in its environment so I can descent into a `qmk_firmware` fork and call `qmk build`. But is this doable? If it starts to compile build tools from ground up then it will be worthless for newcomers.

**Comment by AndersonTorres** [comment]
> But is this doable? If it starts to compile build tools from ground up then it will be worthless for newcomers.

I am not following you here. Are you saying that in every invocation of it, the cross toolchains would be rebuilt from scratch?

As far as I remember, Nix caches the built artifacts at `/nix/store`. It would happen once and never more (except maybe in subsequent GCC updates).

**Comment by bb2020** [comment]
> I am not following you here. Are you saying that in every invocation of it, the cross toolchains would be rebuilt from scratch?
> 
> As far as I remember, Nix caches the built artifacts at `/nix/store`. It would happen once and never more (except maybe in subsequent GCC updates).

Of course, it caches the output. But the last time I tried to use `shell.nix` on my cheap laptop, it tooks days to build gcc :rofl:

File: pkgs/tools/misc/qmk/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/tools/misc/qmk/default.nix:23]: Nope. They are requirements.

**Comment by AndersonTorres** [comment]
Well, in this case the best we can do is ask Hydra maintainers to cache it for those of us with Pentium III machines...

**Comment by AndersonTorres** [comment]
> This can be worked out by overriding package `src`. As long as dependencies are the same, nix package can build their tree too.

But it should be almost the same as rewriting the `qmk clone` functionality.

> It doesn't. We still can't build the firmware with current `qmk_cli` though.

Then it is not the reason why `qmk cli` can't compile the firmwares.

**Comment by bb2020** [comment]
> Then it is not the reason why `qmk cli` can't compile the firmwares.

Dosen't `qmk_cli` just calls `bin/qmk`? 

Shell approach is much better if some packages like `pkgsCross.avr.libcCross` can be cached. 
For example, `gcc-arm-embedded` is cached by Hydra and available instantly in the environment.

**Comment by AndersonTorres** [comment]
> Doesn't `qmk_cli` just calls `bin/qmk`?

https://github.com/qmk/qmk_firmware/blob/master/bin/qmk#L37

> 
> Shell approach is much better if some packages like `pkgsCross.avr.libcCross` can be cached.
> For example, `gcc-arm-embedded` is cached by Hydra and available instantly in the environment.

Yes, and I think we can/should provide it inside Nixpkgs tree.

**Comment by bb2020** [comment]
That would be great. It was always my dream to be able to update my keyboard under NixOS.

File: pkgs/tools/misc/qmk/default.nix (hunk 2 above)
Comment (bhipple) [reply pkgs/tools/misc/qmk/default.nix:23]: We should instead rebase and carry forward the patch from https://github.com/qmk/qmk_cli/pull/48.  Not only are they not real runtime requirements, they're also pinned to exact versions, which is going to be a maintenance headache.
https://github.com/qmk/qmk_cli/blob/master/setup.py#L67

Comment (AndersonTorres) [reply pkgs/tools/misc/qmk/default.nix:23]: > We should instead rebase and carry forward the patch

Can someone do it?

And why this was not patched yet?

**Comment by AndersonTorres** [comment]
Well, that being said, I think it suffices to an official upgrade. We can treat the `qmk-shell.nix` project later.

File: pkgs/tools/misc/qmk/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/qmk/default.nix:23]: We could just remove then with substituteInPlace.

Comment (AndersonTorres) [reply pkgs/tools/misc/qmk/default.nix:23]: A local patch? I am not so sure we should do it if the upstream didn't.

**Comment by AndersonTorres** [comment]
Next phase: `shell.nix` for a full QMK hacking environment!

**Pull request merged**